lib/mocha.js

```diff
@@ -247,6 +258,28 @@ Mocha.prototype._growl = function(runner, reporter) {
 };
 
 /**
+ * Get require.cache files (node only)
+ *
+ * @api private
+ */
+Mocha.prototype._getRequireCache = function() {
+  return Object.keys(require.cache);
+};
+
+/**
+ * Flush test files from require.cache
+ *
+ * @api private
+ */
+Mocha.prototype._flushRequireCache = function() {
+  this._getRequireCache().forEach(function(filePath) {
+    if (this._requireCache.indexOf(filePath) === -1) {
```

Comment:
Since `require()` is imperative, we don't have a guarantee here that the things we're removing from the cache were actually loaded by Mocha--just that they appeared in the cache between *a* and *b*. 

We also don't have a guarantee that everything being removed from the cache is everything that *should* be removed, since your test file could execute a `require()` in async code (use of `--delay` is a more likely scenario).

We could inspect the module's `parent` to see that it was required by `lib/mocha.js`, but we're still going to be unsure about whether module `x` loaded by test `y` should be removed.

Mocha tests are written with the understanding (assumption, more likely) that Mocha doesn't molest `require.cache`.  This change would break assumptions.  I'm also not convinced this should be the default behavior.

Since this is intended to be used programmatically, an option should suffice (a command-line flag probably isn't necessary at this point).